zipline/algorithm.py

```diff
@@ -604,8 +605,7 @@ def _write_and_map_id_index_to_sids(self, identifiers, as_of_date):
             if isinstance(identifier, Asset):
                 asset = self.asset_finder.retrieve_asset(sid=identifier.sid,
                                                          default_none=True)
-
-            elif hasattr(identifier, '__int__'):
+            elif isinstance(identifier, Integral):
```

Comment:
We use `Integral` to define `AssetConvertible` in assets.py, so this seems more in line with the definition there.  

One notable difference is that this will reject `float`s as inputs, which seems more correct to me.  If you pass `3.5` as your index, the existing behavior would truncate to `3` and then convert it to an asset, which seems like unfortunate behavior.